### ISSUE TITLE
[docs] SDK 50 - Android 14 Foreground service permission requirements

### ISSUE THREAD

**Issue by robutler** [expo/expo#26846]
### Summary

With Expo SDK 50 being upgraded to Android SDK 34 (Android 14) there seems to be new foreground permission requirements for submitting an application to Android Play Store.

When trying to submit an application to the Play Store, built with SDK 50 through EAS, I'm getting stopped by the message:
> [logs] Updating track 'internal'...
[logs] Uploading all changes to Google Play...
[logs] Google Api Error: Invalid request - You must let us know whether your app uses any Foreground Service permissions. - Retrying...

My only guess is that this is somehow related to me using `expo-location` in my project, which adds a foreground service to the Android manifest.

From what I can read in Google documentation [here](https://support.google.com/googleplay/android-developer/answer/13392821?sjid=10403373841528529919-EU) there should be a new section in the Play console where I can fill out additional details for Foregrounds services if I'm targeting Android 14 (which SDK 50 does).
However to me it seems like an chicken and the egg situation where I can't even submit my application to let Play console know that I'm targeting that Android version, because the submission process is blocking me, so that new section in the Play console isn't visible.

What is the recommended approach here for an Expo application? The documentation doesn't mention this case from what I can tell.

### Link to the related docs page

https://docs.expo.dev/versions/latest/sdk/location/

**Comment by robutler** [comment]
An update: Doing a manual upload on Play console of the AAB artefact allows me to fill in the new foreground permission details in relation to the new release. I haven't tried submitting to the Play store from EAS again after that, but I suspect that would remove the block from the Google API as well.

**Comment by a-eid** [comment]
> An update: Doing a manual upload on Play console of the AAB artefact allows me to fill in the new foreground permission details in relation to the new release. I haven't tried submitting to the Play store from EAS again after that, but I suspect that would remove the block from the Google API as well. It did make the new "Foreground service permissions" option show up on the Play console under "App content".

what details required to fill in ?

**Comment by robutler** [comment]
> what details required to fill in ?

They want you to declare what foreground permissions you're using, in my case with expo-location it was `android.permission.FOREGROUND_SERVICE_LOCATION`. They also want to know for what purpose you are using the permission. And thirdly attach a link to a video where you demonstrate how you use the permission (I just created an unlisted video on YouTube that I linked).

It's a bit of a hassle, but the review part went quite quickly, within the hour. I don't think they even watched they video according to analytics on YouTube.

**Comment by duggster** [comment]
Same thing happened for me. I got the error on EAS submit, so then I uploaded the AAB manually to the play store. It stopped me and said I had to declare foreground services and provided a link to App Content. It gave a few choices for which permissions my app used (the app asks the user for location on one screen which I use expo-location for), and then they ask for a link to a video showing the functionality in the app. I uploaded an unlisted youtube and put in that link, and that was it.

**Comment by emilnilimaa** [comment]
We get this as well. Tried uploading the aab manually and got the same app content link to fill in.
But when using eas submit for the next version, we get the same error again...
Is it going to be like this for every submit, anyone know?

**Comment by iM-GeeKy** [comment]
Is there any recommendation on what permissions need to be added for [getCurrentPositionAsync()](https://docs.expo.dev/versions/latest/sdk/location/#locationgetcurrentpositionasyncoptions)? My app had been approved by Google, but was rejected yesterday in my latest release that included Expo 50. I have filled out the additional details requested by Google to no avail.

@robutler Thanks for mentioning `android.permission.FOREGROUND_SERVICE_LOCATION`, I did see this in the rejection notes, which I will try to add and resubmit. It would be nice if this was mentioned in the documenation.

**Comment by brteller** [comment]
This kind of seems wrong and should be something we should be able to turn off. The reason being, Google wants this if we're tracking location when the app is not in use, which in my use case, isn't tracked. This is not meant to be for applications that use location when a user has it opened for typical geo located events. Hopefully this can be corrected so we can properly use EAS to submit builds. I'm going to upload a video, but I know what might happen is that they might not like that our app is tracking in the background when it isn't. This also concerns me if this is an iOS issue or not, because they really don't like location tracking in the background either. (Thanks Uber)

**Comment by sato-daiki** [comment]
I'm currently experiencing the same issue. After upgrading to Expo50, I manually uploaded the EAS build file to the Google Play Console. Additionally, I recorded the parts of my app that use maps and uploaded the video to YouTube. However, Google rejected it. Probably because my app doesn't use maps in the background, and the video did not include proof of that.

Currently, with the version upgraded to Expo50, it's not possible to bypass the expo-location issue. Has anyone been able to do it? Is there any good solution?

**Comment by robutler** [comment]
From my understanding even if location is not being tracked in the background, the foreground service permission is still necessary.

See this snippet from the Android documentation:
> If your app contains a feature that shares or receives location information only once, or for a defined amount of time, then that feature requires foreground location access. Some examples include the following:
> Within a navigation app, a feature allows users to get turn-by-turn directions.
> Within a messaging app, a feature allows users to share their current location with another user.

https://developer.android.com/develop/sensors-and-location/location/permissions#foreground

So I don't think removing the foreground service solves that problem. It just seems that Google is really cautious with users sharing position at all.

**Comment by vnil** [comment]
I'm having the same issue where the app got rejected after upgrading to expo 50 because of the added foreground service permission. Like above, I got the error, so I submitted the additional data with a video showing how the location is used (not using it in the background), but got rejected. Would love to hear if anyone got a workaround for this.

**Comment by robutler** [comment]
Just an idea but has anyone tried [blocking](https://docs.expo.dev/versions/latest/config/app/#blockedpermissions) the permission in the app config? I'm guessing Google is only scanning the manifest for this, and if the permission is not there perhaps the app would pass like previously. Of course the would assume you don't need any background location.

**Comment by Ross-Landry** [comment]
> Just an idea but has anyone tried [blocking](https://docs.expo.dev/versions/latest/config/app/#blockedpermissions) the permission in the app config? I'm guessing Google is only scanning the manifest for this, and if the permission is not there perhaps the app would pass like previously. Of course the would assume you don't need any background location.

I had the same idea. Waiting to see the results now, I'll share the outcome in this thread

**Comment by Ross-Landry** [comment]
@robutler That seems to have worked. By blocking the permission in app.config.js I was able to submit the Android build through the API again. 

```
expo : {
 android:{
      "blockedPermissions": [
        "FOREGROUND_SERVICE_LOCATION"
      ]
```

**Comment by vnil** [comment]
@robutler Blocking the permission in the app config did the trick for me, thanks! No error/warning regarding this permission when preparing the release and Google just accepted my new release 🥳, AND the location service worked as expected in the app 😌

**Comment by brentvatne** [comment]
hey folks, we are investigating what the best solution is here that we can introduce as a patch in sdk 50 without breaking anybody's apps. for now, [blocking the foreground service permission](https://github.com/expo/expo/issues/26846#issuecomment-1957291582) is a good workaround. we'll keep you update

**Comment by brentvatne** [comment]
docs are updated, and the foreground service permission is disabled by default in expo-location@16.5.4

**Comment by alterx** [comment]
@brentvatne probably related:

I'm getting this
```
Fatal Exception: java.lang.SecurityException
Starting FGS with type location callerApp=ProcessRecord{6ce728d 8641:xxx.xxxx.xxxxx/u0a382} targetSDK=34 requires permissions: all of the permissions allOf=true [android.permission.FOREGROUND_SERVICE_LOCATION] any of the permissions allOf=false [android.permission.ACCESS_COARSE_LOCATION, android.permission.ACCESS_FINE_LOCATION] and the app must be in the eligible state/exemptions to access the foreground only permission
```
error. Our app hasn't changed and the only thing we did was updating to SDK 50. We don't use the foreground location service explicitly, so the only thing I can think of is that Android 14 (SDK34) included some changes that makes it so the app tries to use it underneath? 

I've tried blocking the permission and I'm on the latest version of `expo-location`. Could it be that now getting the location or just using google maps requires the FOREGROUND_SERVICE_LOCATION permission?

Before this, I didn't have any issues without that permission. App was being built and submitted, and worked as expected.

**Comment by brentvatne** [comment]
@alterx - are you getting this error when attempting to get the current location with google maps? or by calling the expo-location api?

it could be that google maps requires `android.permission.FOREGROUND_SERVICE_LOCATION` in this case. try enabling that with the expo-location config plugin: https://docs.expo.dev/versions/latest/sdk/location/#configurable-properties - `"isAndroidForegroundServiceEnabled": true`

**Comment by alterx** [comment]
@brentvatne  I'm getting at a point in the app where we request the permissions: it calls `getForegroundPermissionsAsync` and requests it if we don't have it. Then calls `getBackgroundPermissionsAsync` and requests it if we don't have it and then sets up a background location task. 

We've never needed the FOREGROUND_SERVICE_LOCATION permission for doing this before

**Comment by alterx** [comment]
@brentvatne yeah, I do get all the permission modals, put after a couple of seconds the app crashes. And upon further investigation, that's the error I'm seeing

**Comment by alterx** [comment]
I was able to remove that part of the code and the app runs just fine, the map also runs fine so it's definitely not it.
I created a new build that calls `getForegroundPermissionsAsync` and then `getCurrentPositionAsync` and I'm able to replicate it so one of those two is probably the culprit @brentvatne

**Comment by brentvatne** [comment]
@alterx - could you possibly create an issue with the full reproducible example an other info? it would be really helpful so we can get someone to quickly investigate this

**Comment by alterx** [comment]
Sorry, I didn't get this notification earlier @brentvatne 
I creaste a new issue here https://github.com/expo/expo/issues/27336

**Comment by emilnilimaa** [comment]
We're getting rejected for foreground service media projection as well...
Can we block this permission too?

**Comment by brentvatne** [comment]
@emilnilimaa - can you provide more information on exactly what you are seeing? we don't add any permissions related to "foreground service media projection" in expo-location

**Comment by emilnilimaa** [comment]
@brentvatne Yeah its probably not expo-location, but we get the same type of rejection. I am not sure which lib it is coming from unfortunately, and we have plenty. But it started appearing at the same time as we received the location rejections...
God damn hate androids :P

**Comment by brentvatne** [comment]
@emilnilimaa - you can search your node_modules for the permission string to see where it's coming from

**Comment by emilnilimaa** [comment]
Which string should we search for?

**Comment by brentvatne** [comment]
@emilnilimaa - whichever string google warned you about :)

**Comment by emilnilimaa** [comment]
Well it only says this:

`Permissions for Foreground Services: Service interruption or deferment of functionality does not have material user impact
Details
We found that one or more of the declared use cases is not compliant with how [foreground service permission](https://support.google.com/googleplay/android-developer/answer/13392821?hl=en&sjid=5928478178588015754-NA) is allowed to be used. Specifically, the declared use case(s) can be interrupted or deferred by the system without creating a negative user experience.
Your app is using foreground service to Media Projection - Other when it is not required to do so.`

None of our code requests any such permission, so it has to be from some lib. But not sure what to search for. Searching for MEDIA_PROJECTION finds nothing.

**Comment by efstathiosntonas** [comment]
@emilnilimaa look for: https://developer.android.com/reference/android/Manifest.permission#FOREGROUND_SERVICE

**Comment by emilnilimaa** [comment]
@efstathiosntonas Doing a search for that in the whole project only gets 1 hit, and thats the in the blockedPermissions I added in app.config.js

**Comment by emilnilimaa** [comment]
In the app bundle explorer in google play i can see these under permissions:
android.permission.FOREGROUND_SERVICE, 
android.permission.FOREGROUND_SERVICE_MEDIA_PROJECTION,
android.permission.FOREGROUND_SERVICE_MICROPHONE,

No idea what puts it in there...

**Comment by emilnilimaa** [comment]
And they seem to go in there even if i add the blockedPermissions in the app.config.js

**Comment by eightyfive** [comment]
I think @brteller is right here. There is no need to enable the permission if you're not using `expo-location` for "background" purposes...

> *TYPE_LOCATION*
>  - Background Location Updates: User-initiated location sharing – For example, Find My Friend feature, vehicle activity tracking.
>  - Background Location Updates: Navigation – For example, continuing driving navigation in maps, ride tracking for ride share.
> - Geofencing – Define parameters, which surround the areas of interest.

https://support.google.com/googleplay/android-developer/answer/13392821?sjid=10403373841528529919-EU

So nothing to do with simple foreground location permissions.

**Comment by radiorapstech** [comment]
Hi, just adding this in case someone's returning here:

**If you're using expo-location, as per the expo-location docs:**

When you install the expo-location module, it automatically adds the following permissions:

ACCESS_COARSE_LOCATION: for approximate device location
ACCESS_FINE_LOCATION: for precise device location
The following permissions are optional:
FOREGROUND_SERVICE and FOREGROUND_SERVICE_LOCATION: to be able to access location while the app is open but backgrounded. FOREGROUND_SERVICE_LOCATION is only required as of Android 14. When you enable this in a new build, you will need to [submit your app for review and request access to use the foreground service permission](https://support.google.com/googleplay/android-developer/answer/13392821?hl=en).
ACCESS_BACKGROUND_LOCATION: to be able to access location while the app is backgrounded or closed. When you enable this in a new build, you will need to [submit your app for review and request access to use the background location permission](https://support.google.com/googleplay/android-developer/answer/9799150?hl=en).

(So indeed, seems like you can exclude FOREGROUND_SERVICE_LOCATION if you don't need background location usage.)

https://github.com/expo/expo/blob/main/docs/pages/versions/unversioned/sdk/location.mdx
